plugins/woocommerce/src/Internal/Traits/SourceAttributionMeta.php

```diff
@@ -0,0 +1,364 @@
+	 */
+	private function filter_meta_data( array $meta ): array {
+		$return = array();
+		$prefix = $this->get_meta_prefixed_field( '' );
+
+		foreach ( $meta as $item ) {
+			if ( str_starts_with( $item->key, $prefix ) ) {
+				$return[ $this->unprefix_meta_field( $item->key ) ] = $item->value;
+			}
+		}
+
+		// Determine the device type from the user agent.
+		if ( ! array_key_exists( 'device_type', $return ) && array_key_exists( 'user_agent', $return ) ) {
+			$return['device_type'] = $this->get_device_type( $return );
+		}
+
+		// Determine the origin based on source type and referrer.
+		$source_type = $return['type'] ?? '';
+		switch ( $source_type ) {
+			case 'organic':
+				$origin = __( 'Organic search', 'woocommerce' );
+				break;
+
+			case 'referral':
+				$origin = __( 'Referral', 'woocommerce' );
+				break;
+
+			case 'typein':
+				$origin = __( 'Direct', 'woocommerce' );
+				break;
+
+			default:
+				$origin = __( 'Unknown', 'woocommerce' );
+				break;
+		}
+
+		$return['origin'] = $origin;
+
+		return $return;
+	}
+
+	/**
+	 * Get the field name with the appropriate prefix.
+	 *
+	 * @param string $field Field name.
+	 *
+	 * @return string The prefixed field name.
+	 */
+	private function get_prefixed_field( $field ): string {
+		return "{$this->field_prefix}{$field}";
+	}
+
+	/**
+	 * Get the field name with the meta prefix.
+	 *
+	 * @param string $field The field name.
+	 *
+	 * @return string The prefixed field name.
+	 */
+	private function get_meta_prefixed_field( string $field ): string {
+		// Map some of the fields to the correct meta name.
+		if ( 'type' === $field ) {
+			$field = 'source_type';
+		} elseif ( 'url' === $field ) {
+			$field = 'referrer';
+		}
+
+		return "_{$this->get_prefixed_field( $field )}";
+	}
+
+	/**
+	 * Remove the meta prefix from the field name.
+	 *
+	 * @param string $field The prefixed field.
+	 *
+	 * @return string
+	 */
+	private function unprefix_meta_field( string $field ): string {
+		$return = str_replace( "_{$this->field_prefix}", '', $field );
+
+		// Map some of the fields to the correct meta name.
+		if ( 'source_type' === $return ) {
+			$return = 'type';
+		} elseif ( 'referrer' === $return ) {
+			$return = 'url';
+		}
+
+		return $return;
+	}
+
+	/**
+	 * Get the order object with HPOS compatibility.
+	 *
+	 * @param WC_Order|WP_Post|int $post_or_order The post ID or object.
+	 *
+	 * @return WC_Order The order object
+	 * @throws Exception When the order isn't found.
+	 */
+	private function get_hpos_order_object( $post_or_order ) {
+		// If we've already got an order object, just return it.
+		if ( $post_or_order instanceof WC_Order ) {
+			return $post_or_order;
+		}
+
+		global $theorder;
+
+		// If we have a post ID, get the post object.
+		if ( is_numeric( $post_or_order ) ) {
+			$post_or_order = get_post( $post_or_order );
+		}
+
+		if ( empty( $theorder ) || $theorder->get_id() !== $post_or_order->ID ) {
+			$theorder = wc_get_order( $post_or_order->ID );
```

Comment:
I don't think a method called `get_...object` should be modifying the global order object. Could we either remove this side effect or give it a different name that makes clear that it has a global side effect beyond just getting an order object?

Reply:
It's not clear why this updates `$theorder`, I've removed the functionality in ac20fafa7406ff6cec07c37352f53c5785e5beff and will continue testing to see if I can see any reason to restore it.